chain/db/pubsub.go

```diff
@@ -2,95 +2,220 @@ package db
+}
+
+func getPubSubId(topic string, bucket uint32, subscriber []byte, identifier string) string {
+	pubSubId := bytes.NewBuffer(nil)
+
+	pubSubId.Write(getTopicBucketId(topic, bucket))
+
+	subscriberHash := sha256.Sum256(append(subscriber, identifier...))
+	pubSubId.Write(subscriberHash[:20])
+
+	return string(pubSubId.Bytes())
+}
+
+func getPubSubCleanupId(height uint32) string {
+	buf := new(bytes.Buffer)
+	_ = serialization.WriteUint32(buf, height)
+	return string(buf.Bytes())
+}
+
+func (sdb *StateDB) getPubSub(id string) (*pubSub, error) {
+	var ps *pubSub
+	var ok bool
+	if ps, ok = sdb.pubSub[id]; !ok {
+		enc, err := sdb.trie.TryGet(append(PubSubPrefix, id...))
+		if err != nil {
+			return nil, err
+		}
+
+		ps = &pubSub{}
+
+		if len(enc) > 0 {
+			buff := bytes.NewBuffer(enc)
+			if err := ps.Deserialize(buff); err != nil {
+				return nil, fmt.Errorf("[getPubSub]Failed to decode state object for pub sub: %v", err)
+			}
+		}
+
+		sdb.pubSub[id] = ps
 	}
 
-	subscriberKey := generateSubscriberKey(subscriber, identifier, topic, bucket)
+	return ps, nil
+}
 
-	// DELETE VALUE
-	err := cs.st.BatchDelete(subscriberKey)
+func (sdb *StateDB) getPubSubCleanup(height uint32) (map[string]struct{}, error) {
+	var psc map[string]struct{}
+	var ok bool
+	if psc, ok = sdb.pubSubCleanup[height]; !ok {
+		enc, err := sdb.trie.TryGet(append(PubSubCleanupPrefix, getPubSubCleanupId(height)...))
+		if err != nil || len(enc) == 0 {
+			return nil, fmt.Errorf("[getPubSubCleanup]can not get pub sub cleanup from trie: %v", err)
+		}
+
+		buff := bytes.NewBuffer(enc)
+		psc = make(map[string]struct{}, 0)
+		pscLength, err := serialization.ReadVarUint(buff, 0)
+		if err != nil {
+			return nil, fmt.Errorf("[getPubSubCleanup]Failed to decode state object for pub sub cleanup: %v", err)
+		}
+		for i := uint64(0); i < pscLength; i++ {
+			id, err := serialization.ReadVarString(buff)
+			if err != nil {
+				return nil, fmt.Errorf("[getPubSubCleanup]Failed to decode state object for pub sub cleanup: %v", err)
+			}
+			psc[id] = struct{}{}
+		}
+
+		sdb.pubSubCleanup[height] = psc
+	}
+
+	return psc, nil
+}
+
+func (sdb *StateDB) cleanupPubSubAtHeight(height uint32, id string) error {
+	ids, err := sdb.getPubSubCleanup(height)
+	if err != nil {
+		return err
+	}
+	ids[id] = struct{}{}
+	return nil
+}
+
+func (sdb *StateDB) cancelPubSubCleanupAtHeight(height uint32, id string) error {
+	ids, err := sdb.getPubSubCleanup(height)
 	if err != nil {
 		return err
 	}
+	if _, ok := ids[id]; ok {
+		delete(ids, id)
+	}
+	return nil
+}
+
+func (sdb *StateDB) subscribe(topic string, bucket uint32, subscriber []byte, identifier string, meta string, expiresAt uint32) error {
+	id := getPubSubId(topic, bucket, subscriber, identifier)
 
-	err = cs.CancelKeyExpirationAtBlock(height+duration, subscriberKey)
+	ps, err := sdb.getPubSub(id)
 	if err != nil {
 		return err
 	}
 
+	if ps.Empty() {
+		ps.subscriber = address.MakeAddressString(subscriber, identifier)
```

Comment:
Do we need to store the full subscriber addr string here? Looks like identifier is enough?

Reply:
why? it's the only place where subscriber is stored, there is no other place where I can get pk from

Reply:
Oh you are right. What about storing identifier (as string) and pk (as []byte) separately to save 32 bytes per subscriber?

Reply:
done